### ISSUE TITLE
Pillar kwargs parse error with #

### ISSUE THREAD

**Issue by dstokes** [saltstack/salt#18045]
The first case causes a `Pillar data must be formatted as a dictionary` error when attempting to execute states.
```shell
$ salt-call --version
salt-call 2014.7.0rc2 (Helium)

$ salt-call test.arg pillar='{"revision": "feature-#1"}'
local:
    ----------
    args:
    kwargs:
        ----------
        __pub_fun:
            test.arg
        __pub_jid:
            20141113005532889179
        __pub_pid:
            12891
        __pub_tgt:
            salt-call
        pillar:
            {"revision": "feature-#1"}

$ salt-call test.arg pillar='{"revision": "feature-1"}'
local:
    ----------
    args:
    kwargs:
        ----------
        __pub_fun:
            test.arg
        __pub_jid:
            20141113005539041122
        __pub_pid:
            12926
        __pub_tgt:
            salt-call
        pillar:
            ----------
            revision:
                feature-1
```

**Comment by JustinCarmony** [comment]
@basepi We were just bit by this because we had a password that had a # in it. So any fix would be awesome. :)

**Comment by basepi** [comment]
I'm still not sure of the source of this bug. However, it appears that YAML itself isn't caught by this:

```
>>> yaml.load('{"revision": "feature-#1"}')
{'revision': 'feature-#1'}
>>> yaml.load('{"revision": "feature-#1"}')['revision']
'feature-#1'
```

So perhaps it is indeed something we're doing. I'll have to find some time to do a dive.

**Comment by basepi** [comment]
Fixed in #24466 

Super easy, once I found the code. xD
